machinelearning/tensorflow.md

```diff
@@ -0,0 +1,349 @@
+```
+aws ec2 run-instances --region us-west-2 --image-id <ami-id from the above command>  --instance-type c7g.4xlarge --count 1  --key-name <key name> --subnet-id <subnet_id> --block-device-mappings 'DeviceName=/dev/sda1,Ebs={VolumeSize=256}'
+```
+
+Once Graviton instance is launched with the above DLAMI, the platform is ready for building Tensorflow based ML inference applications.The DLAMI comes with the additional libraries preinstalled, eg: rust compilers, transformers etc, to enable transformers based (eg: bert) inferencing applications.
+
+
+**Using Docker hub container**
+
+Launch a Graviton instance with an Ubuntu 20.04 AMI and instal docker using the below instructions.
+```
+#install and enable docker
+sudo apt update
+sudo apt install -y docker.io
+sudo systemctl enable docker && sudo systemctl start docker && sudo usermod -a -G docker ubuntu
+sudo chmod 666 /var/run/docker.sock
+
+# pull the tensorflow docker container with onednn-acl optimizations enabled
+docker pull armswdev/tensorflow-arm-neoverse
+
+# launch the docker image
+docker run -it --rm -v /home/ubuntu/:/hostfs armswdev/tensorflow-arm-neoverse
+```
+
+**Using Python wheel**
+Tensorflow official wheel is now available for aarch64, this supports optimized onednn+acl backend for Graviton CPUs.
+```
+pip install aws-tensorflow-cpu
+```
+
+
+# Runtime configurations for optimal performance
+Once the tensorflow setup is ready, enable the below runtime configurations to achieve the best performance.
+
+```
+# The default runtime backend for tensoflow is eigen, but typically onednn+acl provides better performance and this can be enabled by setting the below TF environment variable
+export TF_ENABLE_OENDNN_OPTS=1
+
+# Graviton3 (eg: c7g instances) supports BF16 format for ML acceleration. This can be enabled in oneDNN by setting the below environment variable
+export DNNL_DEFAULT_FPMATH_MODE=BF16
+
+# Tensorflow's inter and intra_op_parallelism_threads and the system openmp thread settings are critical for the optimal workload parallelization in a multi-threaded system.
+# set the inter and intra op thread count during the session creation
+session = Session(
+                 config=ConfigProto(
+                      intra_op_parallelism_threads=<num. of vcpus>,
+                      inter_op_parallelism_threads=1,
+                )
+)
+
+# system openmp environment needs to be set as below
+export OMP_NUM_THREADS = <no. of vcpus> / <no. of procs>
+export OMP_PROC_BIND=false
+export OMP_PLACES=cores
+
+```
+Along with the enviornment setup, it is also required to optimize the graph for inference with tensorflow grappler for constant folding, layout, arithmetic and memory optimizations. For a detailed description about each of the optimizatin passes, please refer to Tensorflow [Grappler](https://www.tensorflow.org/guide/graph_optimization) documentation. Below is a snippet of what libraries to import and how to invoke the Grappler passes for inference.
```

Comment:
required? i doubt it.

Reply:
Yes, this is required to remove the training nodes, fold the batch norms and fuse some of the ops. btw, this is nothing specific for aarch64, I will convey that in the above text to avoid confusion. This is a general recommendation for any model inference on tensorflow irrespective of whether it runs on cpu or gpu or tpu.... 

here is the script for it, part of tensorflow.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference_lib.py